### PR TITLE
Adding of network applet

### PR DESCRIPTION
**Pull request by danielmgit**
- nas: Addition of a copy to restriced dirs functionality and minor fixes
- Network applet added (set direct | indirect connection)
- Implementation of conf parsing in nas (instead of every applet)
- Fixes of bugs
- Code refactoring

### CODE REVIEW

File: conf.json [hunk 1]
```diff
@@ -9,7 +9,9 @@
 				"interfaces": [
 				   {
 						 "name": "enp4s0",
-						 "ip-address": "10.0.0.1"
+					     "type": "test",
+						 "ip-address": "10.0.0.1",
```
Comment (hgn) [inline conf.json]: whitespace error?

File: conf.json [hunk 2]
```diff
@@ -20,7 +22,9 @@
 				"interfaces": [
 				   {
 						 "name": "enp0s25",
-						 "ip-address": "10.0.1.1"
+                         "type": "test",
```
Comment (hgn) [inline conf.json]: whitespace errors? indent correct?

File: nas.py [hunk 3]
```diff
@@ -69,6 +68,21 @@ def copy(self, remote_user, remote_ip, remote_path, local_path, to_local):
         stdout, stderr = p.communicate()
         return stdout, stderr, p.returncode
 
+    def copy_to_restricted(self, user, ip, from_path, to_path,
```
Comment (hgn) [inline nas.py:71]: why not simple copy()?

File: applets/005-netperf/applet.py [hunk 4]
```diff
@@ -8,27 +8,7 @@
 REMOTE_NET_PATH = "/tmp/net-applet-shuffler"
 
 
-def ssh_exec(ip, user, cmd):
```
Comment (hgn) [inline applets/005-netperf/applet.py:11]: is exec() not required anymore?

File: conf.json (hunk 2 above)
Comment (danielmgit) [inline conf.json]: As above, (should be) fixed now.

File: conf.json (hunk 1 above)
Comment (danielmgit) [inline conf.json]: Legacy issues, (should be) fixed now.

File: nas.py (hunk 3 above)
Comment (danielmgit) [inline nas.py:71]: Scp can't copy to user restriced locations directly. The program created sudo /tmp/net-applet-shuffler is not user owned. Thus, either the owner has to be changed, the function in question has to be used.

File: applets/005-netperf/applet.py (hunk 4 above)
Comment (danielmgit) [inline applets/005-netperf/applet.py:11]: The nas exec works fine here. This exec was legacy from daemonizing.

File: applets/005-netperf/applet.py [hunk 5]
```diff
@@ -81,17 +48,18 @@ def main(x, conf, args):
...
+    arg_d["ip_source"] = conf.get_ip(arg_d["name_source"], 0)
+    arg_d["user_source"] = conf.get_user(arg_d["name_source"])
+    arg_d["ip_dest"] = conf.get_ip(arg_d["name_dest"], 0)
+    arg_d["user_dest"] = conf.get_user(arg_d["name_dest"])
 
     # potentially start parallel netperf instances
     # note: at this point, the distribution of the applet will be done, which
     # is probably not the best way to do (can cause unwanted time offsets)
     x.p.msg("Starting netperf applet on host {}\n".format(
                                                         arg_d["user_source"]))
```
Comment (hgn) [inline applets/005-netperf/applet.py:60]: no wrapping required, 80 chars is a **soft**rule.

File: applets/005-netperf/applet.py [hunk 6]
```diff
@@ -81,17 +48,18 @@ def main(x, conf, args):
...
-    contr_thread = Thread(target=controller_thread, args=(arg_d, ))
+    contr_thread = Thread(target=controller_thread, args=(x, arg_d, ))
```
Comment (hgn) [inline applets/005-netperf/applet.py:61]: where is the thread joined? when are the resources are freed? I mean probably not required and implicitly handled within pythons threading library?

File: applets/009-network/applet.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    def __init__(self, user, ip, arguments):
+        Thread.__init__(self)
+        self.daemon = True
+        self.host_user = user
+        self.host_ip = ip
+        self.arguments = arguments
+
+    def run(self):
+        ssh_command = "ssh {}@{} sudo python3.5 {}/{} {}".format(
```
Comment (hgn) [inline applets/009-network/applet.py:29]: should be python3 to be compatible with older and future releases. There is no specific feature required to bound to 3.5.

File: applets/005-netperf/applet.py (hunk 6 above)
Comment (danielmgit) [inline applets/005-netperf/applet.py:61]: The idea at this point is to have non-dependant threads (no joining). This is due to the architecture of nas. Ressources are freed (as far as i understand), as soon as the threads finish. A netperf thread finishes, and thus frees ressources, directly after distribution of the netperf controller and it's startup.

File: applets/009-network/applet.py (hunk 7 above)
Comment (danielmgit) [inline applets/009-network/applet.py:29]: Will be done.

File: nas.py (hunk 3 above)
Comment (hgn) [inline nas.py:71]: Ok, got it. Idea: copy() is now more or less useless. Should we rename copy_to_restricted() to copy and make copy() private? copy() -> _copy()?

The new copy can later be extended by a new argument called **mode** with the options: **restricted** (default) and **non_restricted**.

It makes no sense to show every user copy internals! I mean the user has no gain to know that the function copy is somewhat special and the copy is handled over an indirection.

Comment (danielmgit) [inline nas.py:71]: Sounds definitely like the way to go.

File: applets/005-netperf/applet.py (hunk 6 above)
Comment (hgn) [inline applets/005-netperf/applet.py:61]: ok, thanks! (that is different from standard pthread)

**Pull request merged**